class DomainException(Exception):
    """
    Базовое исключение для ошибок доменной логики.

    Все доменные исключения наследуются от этого класса.
    """
    pass


class AuthenticationException(DomainException):
    """
    Исключение, вызываемое при ошибке аутентификации.

    Возникает, когда аутентификация пользователя не удалась из-за неверных учетных данных.
    """
    pass


class UserAlreadyExistsException(DomainException):
    """
    Исключение, вызываемое при создании пользователя, если пользователь уже существует.

    Возникает при попытке регистрации пользователя с именем, которое уже занято.
    """
    pass


class InvalidTokenException(DomainException):
    """
    Исключение, вызываемое при использовании недействительного токена.

    Возникает, если токен доступа невалиден, истек или не может быть декодирован.
    """
    pass


class ContainerNotFoundException(DomainException):
    """
    Исключение, вызываемое при отсутствии контейнера.

    Возникает, если указанный контейнер не найден в системе.
    """
    pass


class DockerAPIException(DomainException):
    """
    Исключение для ошибок, связанных с API Docker.

    Возникает при ошибках взаимодействия с Docker API, например, при невозможности выполнения команды.
    """
    pass
